ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/QueryResponseHandler.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.api.client.impl;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.datatype.guava.GuavaModule;
+import com.google.common.collect.ImmutableList;
+import io.confluent.ksql.api.server.protocol.QueryResponseMetadata;
+import io.confluent.ksql.util.VertxUtils;
+import io.vertx.core.Context;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.json.jackson.DatabindCodec;
+import io.vertx.core.parsetools.RecordParser;
+import java.util.concurrent.CompletableFuture;
+
+abstract class QueryResponseHandler<T> {
+
+  protected final Context context;
+  protected final RecordParser recordParser;
+  protected final CompletableFuture<T> cf;
+  protected boolean hasReadArguments;
+
+  QueryResponseHandler(final Context context, final RecordParser recordParser,
+      final CompletableFuture<T> cf) {
+    this.context = context;
+    this.recordParser = recordParser;
+    this.cf = cf;
+  }
+
+  public void handleBodyBuffer(final Buffer buff) {
+    checkContext();
+    if (!hasReadArguments) {
+      handleArgs(buff);
+    } else {
+      handleRow(buff);
+    }
+  }
+
+  public void handleBodyEnd(final Void v) {
+    checkContext();
+    handleBodyEnd();
+  }
+
+  protected abstract void handleBodyEnd();
+
+  protected abstract void handleMetadata(final QueryResponseMetadata queryResponseMetadata);
+
+  protected abstract void handleRow(final Buffer buff);
+
+  private void handleArgs(final Buffer buff) {
+    hasReadArguments = true;
+
+    final QueryResponseMetadata queryResponseMetadata;
+    final ObjectMapper objectMapper = DatabindCodec.mapper();
+    objectMapper.registerModule(new GuavaModule());
```

Comment:
Probably don't want to register this every time. And registering modules on the singleton DatabindCodec object mapper will affect any other user of Vert.x in the same process, so probably don't want to do that.
I'd suggest we maintain our own static ObjectMapper in the ClientImpl and use that for all encoding/decoding if possible.

Reply:
Done, though I've put the static ObjectMapper into a separate utility (JsonMapper) rather than in ClientImpl.